torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1652,36 +1625,117 @@ std::string ProcessGroupNCCL::getNCCLWatchdogTimeoutExitMsg(
       ".");
 }
 
-void ProcessGroupNCCL::setEnableNanCheck(bool enableNanCheck) {
-  enableNanCheck_ = enableNanCheck;
+void ProcessGroupNCCL::HeartbeatMonitor::wakeUpMonitor() {
+  monitorWakeUpCV_.notify_one();
+}
+
+bool ProcessGroupNCCL::HeartbeatMonitor::dumpDebuggingInfo(
+    bool includeStackTrace /*=true*/) {
+  return pg_->dumpDebuggingInfo(includeStackTrace);
+}
+
+void ProcessGroupNCCL::HeartbeatMonitor::setLastWorkListUpdateTime(
+    std::chrono::time_point<std::chrono::steady_clock> time) {
+  // We intentially let the race condition to happen but this is ok
+  // as long as we update the time, we know we are making progress.
+  lastWorkListUpdateTime_ = time;
+}
+
+void ProcessGroupNCCL::HeartbeatMonitor::init(int rank, ProcessGroupNCCL* pg) {
+  static c10::once_flag initFlag_;
+  c10::call_once(
+      initFlag_, [&]() { HeartbeatMonitor::get()->doInit(rank, pg); });
+}
+
+void ProcessGroupNCCL::HeartbeatMonitor::doInit(
+    int rank,
+    ProcessGroupNCCL* pg) {
+  pg_ = pg;
+  rank_ = rank;
+  size_ = pg_->getSize();
```

Comment:
we're going to have to get rid of this init function to support multiple PGs right?

Reply:
no the idea is to consolidate this thread as well, so it will work even for multiple PG because the default PG will always be created first. That's the assumption we have for globalRank() as well.